### ISSUE TITLE
Listener count increases when using `useSelector` / StrictMode in React.

### ISSUE THREAD

**Issue by Inuniku** [LegendApp/legend-state#256]
### useSelector

I have following (minimal) react code example using a selector. When dragging the slider, the `listener` count in the `ProxyTarget` of `nestedObservable.value` increases steadily (there seems to be some kind of memory leak) .

[Sandbox](https://codesandbox.io/p/sandbox/funny-raman-w832w7?file=%2Fsrc%2FApp.js%3A28%2C70&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clu8j1got0006356i6hq46hvg%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clu8j1got0002356ih1x0ko6j%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clu8j1got0003356iq95fsc7a%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clu8j1got0005356iysk3pe96%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clu8j1got0002356ih1x0ko6j%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu8j1got0001356iy5runp0j%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clu8jimie0002356hc8ldm176%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A28%252C%2522startColumn%2522%253A70%252C%2522endLineNumber%2522%253A28%252C%2522endColumn%2522%253A70%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clu8j1got0002356ih1x0ko6j%2522%252C%2522activeTabId%2522%253A%2522clu8jimie0002356hc8ldm176%2522%257D%252C%2522clu8j1got0005356iysk3pe96%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu8j1got0004356i448q7d9j%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clu8j1got0005356iysk3pe96%2522%252C%2522activeTabId%2522%253A%2522clu8j1got0004356i448q7d9j%2522%257D%252C%2522clu8j1got0003356iq95fsc7a%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clu8j1got0003356iq95fsc7a%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
```typescript
const nestedObservable = observable({ value: 0 });

function Component() {
    const value = useSelector(nestedObservable.value);

    const onChange: ChangeEventHandler<HTMLInputElement> = (e) => {
        nestedObservable.value.set(+e.target.value);
    };

    return (
        <>
            {String(value)}
            <input type="range" min="1" max="100" value={value} id="myRange" onChange={onChange} />
            <button onClick={() => console.log(nestedObservable.value)}>"Log"</button>
        </>
    );
}
```
This is the output after some slider dragging, when clicking the log button:

> `Proxy(Object) {root: {…}, parent: {…}, key: 'value', proxy: Proxy(Object), listeners: Set(116)} ` <-- increases.

Am I using useSelector wrong? When I'm using the observer(() => Component()) pattern, there is no leak in standard react, but in _strict_ mode:

### Strict Mode

[Sandbox](https://codesandbox.io/p/sandbox/funny-raman-w832w7?file=%2Fsrc%2FApp.js%3A28%2C70&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clu8j1got0006356i6hq46hvg%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clu8j1got0002356ih1x0ko6j%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clu8j1got0003356iq95fsc7a%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clu8j1got0005356iysk3pe96%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clu8j1got0002356ih1x0ko6j%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu8j1got0001356iy5runp0j%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clu8jimie0002356hc8ldm176%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A28%252C%2522startColumn%2522%253A70%252C%2522endLineNumber%2522%253A28%252C%2522endColumn%2522%253A70%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clu8j1got0002356ih1x0ko6j%2522%252C%2522activeTabId%2522%253A%2522clu8jimie0002356hc8ldm176%2522%257D%252C%2522clu8j1got0005356iysk3pe96%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clu8j1got0004356i448q7d9j%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clu8j1got0005356iysk3pe96%2522%252C%2522activeTabId%2522%253A%2522clu8j1got0004356i448q7d9j%2522%257D%252C%2522clu8j1got0003356iq95fsc7a%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clu8j1got0003356iq95fsc7a%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

```typescript
root.render(
    <StrictMode>
        <App />
    </StrictMode>
);

function App() {
    const [checked, setChecked] = useState(false);

    return (
        <div>
            <input type="checkbox" checked={checked} onChange={(e) => setChecked(e.target.checked)} />
            {checked && <Component/> }
        </div>
    );
}

const Component = observer(function Component() {
    const value = nestedObservable.value.get();

    const onChange: ChangeEventHandler<HTMLInputElement> = (e) => {
        nestedObservable.value.set(+e.target.value);
    };

    return (
        <>
            {String(value)}
            <input type="range" min="1" max="100" value={value} id="myRange" onChange={onChange} />
            <button onClick={() => console.log(nestedObservable.value)}>"Log"</button>
        </>
    );
})

```

> `Proxy(Object) {root: {…}, parent: {…}, key: 'value', proxy: Proxy(Object), listeners: Set(16)} ` <-- increases.

When mounting and unmounting the component, the listener counts also increases. Is this the intended behavior?

**Comment by jmeistrich** [comment]
Ah thanks, looks like there was a problem in strict mode. I've pushed a fix (https://github.com/LegendApp/legend-state/commit/5230249bf6b69e8ef24a19462b7915b4ce1ba2d0) but since it's a core change I'll just test it in our apps for a bit to make sure it's safe before releasing an update.

**Comment by Inuniku** [comment]
Thank you for your quick response! This package is surely becoming very helpful for us!

The first case (useSelector) isn't related to strict mode as far as I can tell. Am I using useSelector incorrectly here?

**Comment by jmeistrich** [comment]
That looks correct. Could you try that with the new version of useSelector and see if it fixes it? You could copy the source into your project and see if it works better than the built-in version? https://github.com/LegendApp/legend-state/blob/main/src/react/useSelector.ts

I added some tests which should confirm that the listener count stays at 1 in both strict and non-strict modes. But I wonder if codesandbox or another environment has different behavior than the tests?
https://github.com/LegendApp/legend-state/blob/main/tests/react.test.ts#L322

**Comment by Inuniku** [comment]
@jmeistrich 
Thank your response. Indeed the `useSelector` works flawlessly with a provided selector function. The problem occurs, when inputing a proxy directly like:
```typescript
    const value = useSelector(nestedObservable.value);
``` 
This seems to be supported according to the [docs](https://www.legendapp.com/open-source/state/react/react-api/)

```typescript
const state$ = observable({ selected: 1, theme })

const Component = ({ id }) => {

...
    // Get the raw value of an observable and listen to it
    const theme = useSelector(state$.theme)
}
``` 



This is happening in this provided test case: #257

**Comment by jmeistrich** [comment]
A PR with a failing test is glorious! Thank you for that :)

I just released version 2.1.9 with that fixed. There was an optimization for when the parameter is an observable that I had added long ago, but after recent changes it was double subscribing. That tiny optimization wasn't really important so I just removed it 😅

I think this issue should be all fixed now so I'll close it, but please let me know if you find any other issues. And thanks a ton for all the help!
